### PR TITLE
Gwbse exact fix

### PR DESCRIPTION
**Pull request by JavierSijen**
New branch implementing the exact sigma integration and the grid method for calculating the QP energies.

### CODE REVIEW

File: src/libxtp/gwbse/rpa.cc [hunk 1]
```diff
@@ -81,5 +82,96 @@ Eigen::MatrixXd RPA::calculate_epsilon(double frequency) const {
...
+  Eigen::VectorXd AmB = Calculate_H2p_AmB();
+  Eigen::MatrixXd ApB = Calculate_H2p_ApB();
+
+  Eigen::MatrixXd C = AmB.cwiseSqrt().asDiagonal() * ApB * AmB.cwiseSqrt().asDiagonal();
+  tools::EigenSystem result = Diagonalize_H2p_C(C);
+  
+  // Omega has to have correct size otherwise MKL does not rescale for Sqrt
+  Eigen::VectorXd omega;
+  omega = Eigen::VectorXd::Zero(rpasize);
+  omega = result.eigenvalues().cwiseSqrt();
```
Comment (junghans) [inline src/libxtp/gwbse/rpa.cc]: Codacy Issue found: [Variable 'omega' is reassigned a value before the old one has been used.](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4501065)

Comment (JensWehner) [reply src/libxtp/gwbse/rpa.cc]: yeah merging the first two lines into one should do the trick.

Comment (junghans) [inline src/libxtp/gwbse/rpa.cc]: Codacy Issue found: [Variable 'omega' is reassigned a value before the old one has been used.](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4501065)

Comment (junghans) [inline src/libxtp/gwbse/rpa.cc]: Codacy Issue found: [Variable 'omega' is reassigned a value before the old one has been used.](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4501065)

Comment (junghans) [inline src/libxtp/gwbse/rpa.cc]: Codacy Issue found: [Variable 'omega' is reassigned a value before the old one has been used.](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4501065)

**Comment by JensWehner** [comment]
@JavierSijen get all the tests and builds to pass and then we can review it in more detail. 

advice compile it with gcc, or clang and add `-Wall -Wextra -Wpedantic -Wshadow -Wconversion -Werror` to your Compiler options in CMakeCache in the buil directory, if it passes that you have to fix only a little bit

File: src/libxtp/gwbse/rpa.cc (hunk 1 above)
Comment (JensWehner) [reply src/libxtp/gwbse/rpa.cc]: Just merge the first two lines into one. If that does not kill the Codacy warning, just ignore it.

Comment (junghans) [inline src/libxtp/gwbse/rpa.cc]: Codacy Issue found: [Variable 'omega' is reassigned a value before the old one has been used.](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4501065)

Comment (junghans) [inline src/libxtp/gwbse/rpa.cc]: Codacy Issue found: [Variable 'omega' is reassigned a value before the old one has been used.](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4501065)

Comment (junghans) [inline src/libxtp/gwbse/rpa.cc]: Codacy Issue found: [Variable 'omega' is reassigned a value before the old one has been used.](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4501065)

Comment (junghans) [inline src/libxtp/gwbse/rpa.cc]: Codacy Issue found: [Variable 'omega' is reassigned a value before the old one has been used.](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4501065)

Comment (junghans) [inline src/libxtp/gwbse/rpa.cc]: Codacy Issue found: [Variable 'omega' is reassigned a value before the old one has been used.](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4501065)

File: src/libxtp/gwbse/sigma_exact.cc [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+#include <votca/xtp/rpa.h>
+#include <votca/xtp/sigma_exact.h>
+#include <votca/xtp/threecenter.h>
+
+namespace votca {
+namespace xtp {
+
+void Sigma_Exact::PrepareScreening() {
+  _rpa_solution = _rpa.Diagonalize_H2p();
+  _residues = CalcResidues();
```
Comment (JensWehner) [inline src/libxtp/gwbse/sigma_exact.cc]: Can we reconsider calculating the residues on the fly as they again increase the memory footprint? Where they bigger or smaller than the eigenvectors?

File: include/votca/xtp/gw.h [hunk 3]
```diff
@@ -82,20 +87,23 @@ class GW {
   options _opt;
 
   std::unique_ptr<Sigma_base> _sigma = nullptr;
+  std::unique_ptr<QPGrid> _qpgrid = nullptr;
```
[the remark this answers is not in the document]
Comment (JavierSijen) [reply include/votca/xtp/gw.h]: Grid is no prestage solver atm. It's either grid or fixed-point.

File: src/libxtp/gwbse/qpgrid.cc [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  for (Index i_node = 0; i_node < _opt.steps; ++i_node) {
+    sigmc_mat.col(i_node) =
+        _sigma.CalcCorrelationDiag(frequencies.array() + linSpaced[i_node]);
+  }
+  Eigen::VectorXd roots = frequencies;
+  for (Index level = 0; level < qptotal; ++level) {
+    const double freq_cur = frequencies[level];
+    Eigen::VectorXd grid_freq = freq_cur + linSpaced.array();
+    Eigen::VectorXd grid_sigc = sigmc_mat.row(level);
+    Eigen::VectorXd grid_targ =
```
[the remark this answers is not in the document]
Comment (JavierSijen) [reply src/libxtp/gwbse/qpgrid.cc]: I'd like to keep freq, sigc, targ the same number of characters for nicer notation... The derivative approximations would look really ugly if the variables have different number of characters. But it's up to you of course ;).

File: include/votca/xtp/gw.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (JensWehner) [reply include/votca/xtp/gw.h]: okay works for the moment

**Pull request merged**